### PR TITLE
Fixes navigator.credentials.preventSilentAccess()

### PR DESCRIPTION
**Pull request by lukewarlow**
#### 40294715d0c0e48bf27775eadfe7b76ec2e62250
<pre>
Fixes navigator.credentials.preventSilentAccess()
<a href="https://bugs.webkit.org/show_bug.cgi?id=242002">https://bugs.webkit.org/show_bug.cgi?id=242002</a>

Reviewed by J Pascoe.

This change makes the function resolve rather than throw matching the spec.

* LayoutTests/http/wpt/credential-management/credentialscontainer-preventSilentAccess-basics.https.html:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
 (WebCore::CredentialsContainer::preventSilentAccess const):

Canonical link: <a href="https://commits.webkit.org/264345@main">https://commits.webkit.org/264345@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0c40228d830a20bda54f3bef2f6195ba14dd5643

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/661 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3808 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12318 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3586 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105068 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/437 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45704 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98692 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/93317 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13589 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9876 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14265 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52466 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1757 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19637 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16057 "Built successfully") | | |

### CODE REVIEW

**Review by cdumez — CHANGES_REQUESTED** [review]
How is this correct? Now we resolve the promise, making it look like we actually support `preventSilentAccess()`. However, our implementation does nothing. Unless the spec says that calling this function should do nothing besides resolving the promise, I don't think the claim in your commit log is true.

**Review by cdumez — COMMENTED** [review]
Please clarify.

**Comment by lukewarlow** [comment]
Currently Safari exposes a function that unconditionally throws which also makes it look like it supports functionality that it doesn't. This change matches Firefox behaviour (they also don't Implement this functionality) so is mostly an interop fix. The current behaviour breaks feature detection meaning you have to also try catch the code.

No where in the specification does it allow for throwing a not implemented exception. The algorithm from my reading does allow for the function to have no effect though which is what I mean by match the spec.

The alternative fix would be to no longer web expose this function which would also fix feature detection I'm happy with either solution so if you'd prefer that I can change this PR (idk quite how to go about not exposing the function)

**Comment by lukewarlow** [comment]
@cdumez hopefully clarified above

**Comment by cdumez** [comment]
> Currently Safari exposes a function that unconditionally throws which also makes it look like it supports functionality that it doesn't. This change matches Firefox behaviour (they also don't Implement this functionality) so is mostly an interop fix. The current behaviour breaks feature detection meaning you have to also try catch the code.

We throw with a NotSupportedError, which makes feature detection really easy. It is also pretty common that we add stubs for functions have have them reject promises in such way.

By doing nothing and resolving the promise, we make it look like we support this operation, even though we do not. This, in my opinion, actually does break feature detection.

I don't work in this area of the code so I'll defer to @pascoej (who I think may be the current maintainer). My personal preference would be that we add proper support for this (though I realize it is more work).

**Comment by cdumez** [comment]
> > We throw with a NotSupportedError, which makes feature detection really easy. It is also pretty common that we add stubs for functions have have them reject promises in such way.
> 
> As a web developer I expect a web exposed function to be implemented. So I can check for its existence before executing the code.
> 
> > By doing nothing and resolving the promise, we make it look like we support this operation, even though we do not. This, in my opinion, actually does break feature detection.
> 
> By calling this code I expect the browser to not allow silent access to the credentials, this only applies afaik to password credentials which Safari doesn't support. So by calling this function and having it resolve, it works exactly as expected it no longer allows silent access (because it in fact doesn't allow any access). Firefox behaves like this currently.
> 
> > I don't work in this area of the code so I'll defer to @pascoej (who I think may be the current maintainer). My personal preference would be that we add proper support for this (though I realize it is more work).

Ok. I think this makes sense but I'll let @pascoej do the formal review.
Thanks for clarifying. I agree that ideally we wouldn't have exposed an API we don't implement. However, since it was exposed years ago, it would probably be riskier to stop exposing it now.

> I personally would prefer a full password credential implementation too but I assumed WebKit was against this idea hence not implementing it to start with?

No idea what the story behind this is. I assumed we simply hadn't gotten around to implementing it yet.

**Review by pascoej — APPROVED** [review]

**Pull request merged**
